File: .old/Aclass_Python/A1155.py
n = int(input())
num = list(map(int, input().split()))
min_flag = True
max_flag = True
ans = []
index = 1
def compare(a, b):
    global min_flag, max_flag
    if num[a] > num[b]:
        min_flag = False
    elif num[a] < num[b]:
        max_flag = False

def dfs(index, path):
    if index * 2 + 1 >= n and index * 2 + 2 >= n:
        ans.append(path.copy())
        return
    if index * 2 + 2 < n:
        compare(index, index * 2 + 2)
        path.append(num[index * 2 + 2])
        dfs(index * 2 + 2, path)
        path.pop()
    if index * 2 + 1 < n:
        compare(index, index * 2 + 1)
        path.append(num[index * 2 + 1])
        dfs(index * 2 + 1, path)
        path.pop()

temp_path = [num[0]]
dfs(0, temp_path)
for path in ans:
    print(" ".join(list(map(str, path))))
if min_flag:
    print("Min Heap")
elif max_flag:
    print("Max Heap")
else:
    print("Not Heap")

File: .old/Aclass_one_more_time/A1155.cpp
#include <iostream>
#include <vector>
using namespace std;
int min_flag = 1;
int max_flag = 1;
int n;
vector<vector<int> > ans;
vector<int> num;
vector<int> path;
void compare(int a, int b){
    if (a > b) min_flag = 0;
    else if(a < b) max_flag = 0;
}

void dfs(int index){
    if(index * 2 + 2 < n) {
        int next_index = index * 2 + 2;
        path.push_back(num[next_index]);
        compare(num[index], num[next_index]);
        dfs(next_index);
        //path.pop_back();
    }
    if(index * 2 + 1 < n) {
        int next_index = index * 2 + 1;
        path.push_back(num[next_index]);
        compare(num[index], num[next_index]);
        dfs(next_index);
        path.pop_back();
    } else {
        ans.push_back(path);
        path.pop_back();
    }
}

int main(){
    cin >> n;
    num.resize(n);
    for(int i = 0; i < n; i++){
        cin >> num[i];
    }
    path.push_back(num[0]);
    dfs(0);
    for(int i = 0; i < ans.size(); i++){
        for(int j = 0; j < ans[i].size(); j++){
            if(j != 0) cout << " " << ans[i][j];
            else cout << ans[i][j];
        }
        cout << endl;
    }
    if(min_flag == 1) cout << "Min Heap" << endl;
    else if(max_flag == 1) cout << "Max Heap" << endl;
    else cout << "Not Heap" << endl;
    return 0;
}